### ISSUE TITLE
implement SeaThru-NeRF in your own NeRF

### ISSUE THREAD

**Issue by MMMazart** [deborahLevy130/seathru_NeRF#9]
H, thank you for your excellent work, I have implemented SeaThru-NeRF in my own NeRF following your tutorial, but during training, the model tends to fit the whole scene with C_obj and the values of C_med are all 0. How do you avoid this? Looking forward to your reply, thanks!

**Comment by MMMazart** [comment]
I have observed that in your code, this is done very well, bs will not approach 0, but it seems that you have not added additional constraints, how do you do this? Looking forward to your reply！thanks！

**Comment by deborahLevy130** [comment]
Hi,
I need more details about your implementation please.

**Comment by MMMazart** [comment]
Hi,
thank you for your reply!
I modified the vanilla nerf code.
1. used recon loss instead of mse loss;
`def recon_loss(rgb,target_s):
    rgb_render_clip = torch.minimum(torch.tensor(1.), rgb)
    resid_sq_clip = (rgb_render_clip - target_s[..., :3]) ** 2
    scaling_grad = 1. / (1e-3 + rgb_render_clip.detach())
    data_loss = resid_sq_clip * scaling_grad ** 2
    return data_loss.mean()`
2. used extra mlp to predict sigma_haze as in the paper, netdepth = 6 , netwidth = 256;
`        h = input_views_haze
        for i, l in enumerate(self.pts_linears_haze):
            h = self.pts_linears_haze[i](h)
            h = F.softplus(h)
            if i in self.skips:   
                h = torch.cat([input_views_haze, h], -1)

        sigma = self.sigma_linear(h)
        rgb = self.rgb_linear_haze(h)

        outputs_haze = torch.cat([rgb, sigma], -1)
        return outputs,outputs_haze   `
3. implement volume rendering according to the formula in your paper，referred to your code
`
    sigma2alpha = lambda sigma, dists: 1.-torch.exp(-sigma*dists)

    dists = z_vals[...,1:] - z_vals[...,:-1]
    t_vals = z_vals*torch.norm(rays_d[...,None,:], dim=-1)
    dists = torch.cat([dists, torch.Tensor([1e10]).expand(dists[...,:1].shape)], -1)  # [N_rays, N_samples] 

    dists = dists * torch.norm(rays_d[...,None,:], dim=-1)  
                    
    rgb_c = torch.sigmoid(raw_c[...,:3])  # [N_rays, N_samples, 3]
    
    rgb_h = torch.sigmoid(raw_h[...,:3]) # [N_rays, 3]

    noise = 0.
    noise_haze = 0.

    if raw_noise_std > 0.:
        noise = torch.randn(raw_c[...,3].shape) * raw_noise_std
        noise_haze = torch.randn(raw_h[...,3].shape) * raw_noise_std
        # Overwrite randomly sampled data if pytest
        if pytest:
            np.random.seed(0)
            noise = np.random.rand(*list(raw_c[...,3].shape)) * raw_noise_std
            noise = torch.Tensor(noise)


    sigma_c = F.softplus(raw_c[...,3] + noise)
    sigma_h = F.softplus(raw_h[...,3] + noise_haze)  #[1024]

    sigma_h = sigma_h[:,None]            #[1024,1]

    trans_h = torch.exp(-sigma_h*z_vals*torch.norm(rays_d[...,None,:], dim=-1))

    alpha_c = sigma2alpha(sigma_c, dists) 
    alpha_h = sigma2alpha(sigma_h, dists) 

    t = torch.cat([torch.ones((alpha_c.shape[0], 1)), 1.-alpha_c + 1e-10], -1)
    T = torch.cumprod(t, -1)[:, :-1]

    weights_c =  T  * alpha_c

    weights_h =  trans_h * alpha_h

    J = torch.sum(weights_c[...,None] * rgb_c, -2)

    rgb_c =  torch.sum(trans_h[...,None] * weights_c[...,None] * rgb_c, -2)  # [N_rays, 3]

    rgb_h =  torch.sum(T[...,None] * weights_h[...,None] * rgb_h[:,None,:], -2)  # [N_rays, 3]

    rgb = rgb_c + rgb_h
`
during training, the model tends to fit the whole scene with C_obj and the values of C_med are all 0.
Looking forwarding to your reply, thank you very much!

**Comment by MMMazart** [comment]
Hi,
thank you for your reply!
I modified the vanilla nerf code.
1. used recon loss instead of mse loss;
```
def recon_loss(rgb,target_s):
    rgb_render_clip = torch.minimum(torch.tensor(1.), rgb)
    resid_sq_clip = (rgb_render_clip - target_s[..., :3]) ** 2
    scaling_grad = 1. / (1e-3 + rgb_render_clip.detach())
    data_loss = resid_sq_clip * scaling_grad ** 2
    return data_loss.mean()
```
2. used extra mlp to predict sigma_haze as in the paper, netdepth = 6 , netwidth = 256;
```
h = input_views_haze
for i, l in enumerate(self.pts_linears_haze):
    h = self.pts_linears_haze[i](h)
    h = F.softplus(h)
    if i in self.skips:   
        h = torch.cat([input_views_haze, h], -1)

sigma = self.sigma_linear(h)
rgb = self.rgb_linear_haze(h)

outputs_haze = torch.cat([rgb, sigma], -1)
return outputs,outputs_haze  
```
3. implement volume rendering according to the formula in your paper，referred to your code
```
sigma2alpha = lambda sigma, dists: 1.-torch.exp(-sigma*dists)

dists = z_vals[...,1:] - z_vals[...,:-1]
t_vals = z_vals*torch.norm(rays_d[...,None,:], dim=-1)
dists = torch.cat([dists, torch.Tensor([1e10]).expand(dists[...,:1].shape)], -1)  # [N_rays, N_samples] 

dists = dists * torch.norm(rays_d[...,None,:], dim=-1)  
              
rgb_c = torch.sigmoid(raw_c[...,:3])  # [N_rays, N_samples, 3]

rgb_h = torch.sigmoid(raw_h[...,:3]) # [N_rays, 3]

noise = 0.
noise_haze = 0.

if raw_noise_std > 0.:
  noise = torch.randn(raw_c[...,3].shape) * raw_noise_std
  noise_haze = torch.randn(raw_h[...,3].shape) * raw_noise_std
  # Overwrite randomly sampled data if pytest
  if pytest:
      np.random.seed(0)
      noise = np.random.rand(*list(raw_c[...,3].shape)) * raw_noise_std
      noise = torch.Tensor(noise)


sigma_c = F.softplus(raw_c[...,3] + noise)
sigma_h = F.softplus(raw_h[...,3] + noise_haze)  #[1024]

sigma_h = sigma_h[:,None]            #[1024,1]

trans_h = torch.exp(-sigma_h*z_vals*torch.norm(rays_d[...,None,:], dim=-1))

alpha_c = sigma2alpha(sigma_c, dists) 
alpha_h = sigma2alpha(sigma_h, dists) 

t = torch.cat([torch.ones((alpha_c.shape[0], 1)), 1.-alpha_c + 1e-10], -1)
T = torch.cumprod(t, -1)[:, :-1]

weights_c =  T  * alpha_c

weights_h =  trans_h * alpha_h

J = torch.sum(weights_c[...,None] * rgb_c, -2)

rgb_c =  torch.sum(trans_h[...,None] * weights_c[...,None] * rgb_c, -2)  # [N_rays, 3]

rgb_h =  torch.sum(T[...,None] * weights_h[...,None] * rgb_h[:,None,:], -2)  # [N_rays, 3]

rgb = rgb_c + rgb_h
```
during training, the model tends to fit the whole scene with C_obj and the values of C_med areGradually converging to 0.
Looking forwarding to your reply, thank you very much!

**Comment by lkk123841** [comment]
Hello, may I ask if I can consult with you on how to implement SeaThru-NeRF step by step on top of the original NeRF? I am currently troubled by this issue and look forward to your response. Thank you!
